src/monitormanager.cpp

```diff
@@ -213,6 +214,71 @@ void MonitorManager::removeMonitor(HSMonitor* monitor)
     monitor_update_focus_objects();
 }
 
+int MonitorManager::addMonitor(Input input, Output output)
+{
+    // usage: add_monitor RECTANGLE [TAG [NAME]]
+    if (input.size() < 2) {
+        return HERBST_NEED_MORE_ARGS;
+    }
+
+    input.shift();
+    auto rect = Rectangle::fromStr(input.front());
+
+    input.shift();
+    HSTag* tag;
+    if (input.empty()) {
+        tag = find_unused_tag();
+        if (!tag) {
+            output << input.command() << ": There are not enough free tags\n";
+            return HERBST_TAG_IN_USE;
+        }
+    } else {
+        tag = find_tag(input.front().c_str());
+        if (!tag) {
+            output << input.command() << ": Tag \"" << input.front() << "\" does not exist\n";
+            return HERBST_INVALID_ARGUMENT;
+        }
+        if (find_monitor_with_tag(tag)) {
+            output << input.command() <<
+                ": Tag \"" << input.front() << "\" is already being viewed on a monitor\n";
+            return HERBST_TAG_IN_USE;
+        }
+    }
+
+    input.shift();
+    std::string name;
+    if (!input.empty()) {
+        name = input.front();
```

Comment:
I think you can even improve this by writing:
```
if (input.read({ &name })) {
```
And similarly earlier where the rectangle and the tag name is read.

Reply:
I really think this is a bad interface (from a C++ point of view) but it serves is purpose to extract several arguments at once. For the single argument case I already intended to propose a `pop()` method that combines front with shift.

Reply:
I've changed it for the monitor name, but it wasn't immediately clear to me how to use `read()` for the rect and tag without complicating things with additional local variables. Am I missing something?